src/main/java/com/vackosar/gitflowincrementalbuild/boundary/MavenLifecycleParticipant.java

```diff
@@ -68,6 +76,22 @@ public void afterProjectsRead(MavenSession session) throws MavenExecutionExcepti
         logger.info("gitflow-incremental-builder exiting...");
     }
 
+    private boolean isEnabledForBranch(MavenSession session) throws IOException {
+        if (configProvider.get().enabledBranchRegex == null) {
+            return true;
+        }
+        
+        try (GitFactory gitFactory = createGitFactory(session, configProvider.get())) {
```

Comment:
This will lead to two initializations of JGit in total (in case `enabledBranchRegex` is set and does not match) which means unnecessary overhead.
We have to avoid that.

One solution would be to defer the closing of `Git` to the end of `MavenLifecycleParticipant`.

I can't say I like having JGit usage scattered over two very different parts of GIB but right now I have not good alternate suggestion...

`UnchangedProjectsRemover` could do the skipping but that would require `DifferentFiles` to do the branch check and to expose it to `ChangedProjects` which would need to pass it through to `UnchangedProjectsRemover`.
Also far from ideal...

Reply:
Based on my limited experience with JGit, the initialization does not seem to be expensive.

I changed the code so that initialization is done only once by maintaining a thread local reference. Far from ideal too but please let me know if you prefer something different.